File: src/canvas/curtainPanel.ts
// ────────────────────────────────────────────────────────────────────────────
// MEMB-3 — Rideau : le panneau personnel qui se pose sur une membrane focalisée.
//
// Un rideau n'existe qu'en mode Focus (cf. `membraneFocus`) et nulle part
// ailleurs. Il est collé au BORD DROIT, façon console Quake ou Slide Over :
// il RECOUVRE le canvas focalisé, qui lui ne bouge jamais. La seule grandeur
// qui varie est la largeur du panneau.
//
//   replié   — une languette (≈ 1/10 de l'écran) reste visible à droite ;
//   déployé  — le panneau prend l'essentiel de l'écran, et il reste une bande
//              de canvas à gauche.
//
// Les DEUX sont donc toujours visibles, quel que soit l'état. C'est la règle
// qui borne les réglages : ni la languette ni la bande de canvas ne peuvent
// tomber à zéro (cf. `normalizeConfig`).
//
// Le basculement se fait au SURVOL, sans clic : la souris entre sur le panneau,
// il se déploie ; elle retourne sur le canvas, il se replie.
//
// POURQUOI CE SURVOL NE PEUT PAS BATTRE. Un survol qui redimensionne la zone
// qu'on survole est le cas d'école du clignotement : la frontière bouge, passe
// sous le curseur, inverse la décision, revient… Ici c'est impossible, et pas
// par réglage fin — par géométrie :
//
//   • se déployer déplace la frontière vers la GAUCHE. Un curseur qui était à
//     droite d'elle (donc sur le panneau, ce qui a déclenché le déploiement) y
//     reste, puisqu'elle s'en éloigne.
//   • se replier déplace la frontière vers la DROITE. Un curseur qui était à
//     gauche d'elle (donc sur le canvas) y reste, pour la même raison.
//
// L'animation RENFORCE donc toujours la condition qui l'a déclenchée. Le
// mouvement est auto-stabilisant, et `isOverPanel` le vérifie sur toutes les
// positions intermédiaires. Une temporisation s'ajoute par-dessus, non pas
// contre le battement mais contre l'ouverture accidentelle : la souris traverse
// souvent le bord droit de l'écran en travaillant.
// ────────────────────────────────────────────────────────────────────────────

export const CURTAIN = {
  /** Largeur du rideau replié, en fraction de l'écran — la « languette ». */
  DEFAULT_COLLAPSED: 0.1,
  /** Largeur une fois déployé ; le reste laisse voir le canvas focalisé. */
  DEFAULT_EXPANDED: 0.9,

  // Bornes de réglage. Elles garantissent l'invariant « les deux surfaces
  // restent visibles » : la languette ne disparaît jamais, la bande de canvas
  // non plus.
  MIN_COLLAPSED: 0.02,
  MAX_COLLAPSED: 0.3,
  MIN_EXPANDED: 0.3,
  MAX_EXPANDED: 0.97,
  /** Écart minimal entre replié et déployé — sinon le geste n'a plus de sens. */
  MIN_SPAN: 0.05,

  /** Temporisation avant déploiement. Contre l'ouverture accidentelle en
   *  frôlant le bord droit, pas contre le battement (impossible par géométrie). */
  EXPAND_DWELL_MS: 90,
  /** Repli plus vif : on veut récupérer son canvas sans attendre. */
  COLLAPSE_DWELL_MS: 40,
  /** Constante de temps du glissement. Lissage exponentiel : jamais de
   *  dépassement, et indépendant de la cadence d'images. */
  ANIM_MS: 260,
} as const;

export type CurtainPhase = "collapsed" | "expanded";

export interface CurtainConfig {
  /** Fraction d'écran occupée replié. */
  collapsed: number;
  /** Fraction d'écran occupée déployé. */
  expanded: number;
}

export interface CurtainState {
  /** Fraction d'écran occupée MAINTENANT (animée entre les deux réglages). */
  ratio: number;
  /** État visé. */
  phase: CurtainPhase;
  /** Bascule réclamée par le survol, en attente de temporisation. */
  pending: CurtainPhase | null;
  /** Instant où cette réclamation a commencé. */
  pendingSince: number;
}

export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

/**
 * Réglages valides, quoi qu'on lui donne.
 *
 * Garantit l'invariant du module : `0 < collapsed < expanded < 1`. Les deux
 * surfaces restent donc toujours visibles, y compris si un projet arrive avec
 * des valeurs aberrantes (fichier édité à la main, version future, bug amont).
 */
export function normalizeConfig(cfg?: Partial<CurtainConfig> | null): CurtainConfig {
  const collapsed = clamp(
    cfg?.collapsed ?? CURTAIN.DEFAULT_COLLAPSED,
    CURTAIN.MIN_COLLAPSED,
    CURTAIN.MAX_COLLAPSED,
  );
  const expanded = clamp(
    cfg?.expanded ?? CURTAIN.DEFAULT_EXPANDED,
    Math.max(CURTAIN.MIN_EXPANDED, collapsed + CURTAIN.MIN_SPAN),
    CURTAIN.MAX_EXPANDED,
  );
  return { collapsed, expanded };
}

function clamp(v: number, lo: number, hi: number): number {
  if (!Number.isFinite(v)) return lo;
  return Math.min(hi, Math.max(lo, v));
}

/** État initial : replié. */
export function initialState(cfg: CurtainConfig): CurtainState {
  return { ratio: cfg.collapsed, phase: "collapsed", pending: null, pendingSince: 0 };
}

// ════════════════════════════════════════════════════════════════════════════
// Géométrie
// ════════════════════════════════════════════════════════════════════════════

/** Rectangle du panneau, ancré au bord DROIT. */
export function panelRect(ratio: number, screen: { width: number; height: number }): Rect {
  const width = screen.width * ratio;
  return { x: screen.width - width, y: 0, width, height: screen.height };
}

/**
 * Bande de canvas focalisé restée visible à gauche du panneau.
 *
 * Définie comme le COMPLÉMENT EXACT du panneau, et non par `width * (1 - ratio)` :
 * en virgule flottante les deux ne se rejoignent pas toujours (1000 × (1 − 0,9)
 * vaut 99,999…) et il resterait un liseré d'un pixel entre les deux surfaces.
 */
export function canvasStrip(ratio: number, screen: { width: number; height: number }): Rect {
  return { x: 0, y: 0, width: screen.width - panelRect(ratio, screen).width, height: screen.height };
}

/** Le curseur est-il sur le panneau ? `null` = souris hors fenêtre. */
export function isOverPanel(
  cursorX: number | null,
  ratio: number,
  screen: { width: number },
): boolean {
  if (cursorX === null) return false;
  return cursorX >= screen.width * (1 - ratio);
}

// ════════════════════════════════════════════════════════════════════════════
// Décision & animation
// ════════════════════════════════════════════════════════════════════════════

/**
 * Met à jour la phase visée d'après la position du curseur, temporisation
 * comprise. N'anime rien — c'est `advance` qui déplace le panneau.
 *
 * Souris hors fenêtre (`cursorX === null`) : on se replie. Le rideau ne doit
 * pas rester ouvert derrière une fenêtre qu'on vient de quitter.
 */
export function decide(
  state: CurtainState,
  cursorX: number | null,
  screen: { width: number },
  now: number,
): CurtainState {
  const want: CurtainPhase = isOverPanel(cursorX, state.ratio, screen) ? "expanded" : "collapsed";

  if (want === state.phase) {
    return state.pending === null ? state : { ...state, pending: null, pendingSince: 0 };
  }

  if (state.pending !== want) {
    return { ...state, pending: want, pendingSince: now };
  }

  const dwell = want === "expanded" ? CURTAIN.EXPAND_DWELL_MS : CURTAIN.COLLAPSE_DWELL_MS;
  if (now - state.pendingSince < dwell) return state;

  return { ...state, phase: want, pending: null, pendingSince: 0 };
}

/**
 * Rapproche le panneau de sa cible. Lissage exponentiel : monotone, sans
 * dépassement, et le pas ne dépend pas de la cadence d'images — une image
 * sautée ne fait pas sauter le rideau.
 */
export function advance(state: CurtainState, cfg: CurtainConfig, dtMs: number): CurtainState {
  const target = state.phase === "expanded" ? cfg.expanded : cfg.collapsed;
  if (state.ratio === target) return state;

  const tau = CURTAIN.ANIM_MS / 3;
  const k = 1 - Math.exp(-Math.max(0, dtMs) / tau);
  let ratio = state.ratio + (target - state.ratio) * k;
  // Fin de course : on colle à la cible plutôt que de l'approcher indéfiniment.
  if (Math.abs(target - ratio) < 0.0005) ratio = target;
  return { ...state, ratio };
}

/** Un pas complet : décider, puis animer. */
export function step(
  state: CurtainState,
  cursorX: number | null,
  screen: { width: number },
  cfg: CurtainConfig,
  now: number,
  dtMs: number,
): CurtainState {
  return advance(decide(state, cursorX, screen, now), cfg, dtMs);
}

File: crates/glucose-core/src/curtain_panel.rs
//! MEMB-3 — Rideau : le panneau personnel (géométrie et temporisation).
//! 100% Rust Standard Library (0 dépendance).

use crate::geometry::Rect;

pub mod curtain_consts {
    pub const DEFAULT_COLLAPSED: f64 = 0.1;
    pub const DEFAULT_EXPANDED: f64 = 0.9;
    pub const MIN_COLLAPSED: f64 = 0.02;
    pub const MAX_COLLAPSED: f64 = 0.3;
    pub const MIN_EXPANDED: f64 = 0.3;
    pub const MAX_EXPANDED: f64 = 0.97;
    pub const MIN_SPAN: f64 = 0.05;

    pub const EXPAND_DWELL_MS: i64 = 90;
    pub const COLLAPSE_DWELL_MS: i64 = 40;
    pub const ANIM_MS: i64 = 260;
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CurtainPhase {
    Collapsed,
    Expanded,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CurtainConfig {
    pub collapsed: f64,
    pub expanded: f64,
}

impl Default for CurtainConfig {
    fn default() -> Self {
        Self {
            collapsed: curtain_consts::DEFAULT_COLLAPSED,
            expanded: curtain_consts::DEFAULT_EXPANDED,
        }
    }
}

fn clamp(v: f64, lo: f64, hi: f64) -> f64 {
    if !v.is_finite() {
        lo
    } else {
        v.max(lo).min(hi)
    }
}

pub fn normalize_config(cfg: Option<CurtainConfig>) -> CurtainConfig {
    let raw = cfg.unwrap_or_default();
    let collapsed = clamp(
        raw.collapsed,
        curtain_consts::MIN_COLLAPSED,
        curtain_consts::MAX_COLLAPSED,
    );
    let expanded = clamp(
        raw.expanded,
        curtain_consts::MIN_EXPANDED.max(collapsed + curtain_consts::MIN_SPAN),
        curtain_consts::MAX_EXPANDED,
    );
    CurtainConfig {
        collapsed,
        expanded,
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct CurtainState {
    pub ratio: f64,
    pub phase: CurtainPhase,
    pub pending: Option<CurtainPhase>,
    pub pending_since: i64,
}

impl CurtainState {
    pub fn new(cfg: CurtainConfig) -> Self {
        Self {
            ratio: cfg.collapsed,
            phase: CurtainPhase::Collapsed,
            pending: None,
            pending_since: 0,
        }
    }
}

pub fn panel_rect(ratio: f64, screen_w: f64, screen_h: f64) -> Rect {
    let width = screen_w * ratio;
    Rect::new(screen_w - width, 0.0, width, screen_h)
}

pub fn canvas_strip(ratio: f64, screen_w: f64, screen_h: f64) -> Rect {
    let p = panel_rect(ratio, screen_w, screen_h);
    Rect::new(0.0, 0.0, screen_w - p.width, screen_h)
}

pub fn is_over_panel(cursor_x: Option<f64>, ratio: f64, screen_w: f64) -> bool {
    match cursor_x {
        Some(cx) => cx >= screen_w * (1.0 - ratio),
        None => false,
    }
}

pub fn decide(
    state: &CurtainState,
    cursor_x: Option<f64>,
    screen_w: f64,
    now: i64,
) -> CurtainState {
    let want = if is_over_panel(cursor_x, state.ratio, screen_w) {
        CurtainPhase::Expanded
    } else {
        CurtainPhase::Collapsed
    };

    if want == state.phase {
        if state.pending.is_none() {
            return state.clone();
        } else {
            let mut s = state.clone();
            s.pending = None;
            s.pending_since = 0;
            return s;
        }
    }

    if state.pending != Some(want) {
        let mut s = state.clone();
        s.pending = Some(want);
        s.pending_since = now;
        return s;
    }

    let dwell = if want == CurtainPhase::Expanded {
        curtain_consts::EXPAND_DWELL_MS
    } else {
        curtain_consts::COLLAPSE_DWELL_MS
    };

    if now - state.pending_since < dwell {
        return state.clone();
    }

    let mut s = state.clone();
    s.phase = want;
    s.pending = None;
    s.pending_since = 0;
    s
}

pub fn advance(state: &CurtainState, cfg: &CurtainConfig, dt_ms: f64) -> CurtainState {
    let target = match state.phase {
        CurtainPhase::Expanded => cfg.expanded,
        CurtainPhase::Collapsed => cfg.collapsed,
    };
    if (state.ratio - target).abs() < 1e-9 {
        return state.clone();
    }

    let tau = (curtain_consts::ANIM_MS as f64) / 3.0;
    let k = 1.0 - (-dt_ms.max(0.0) / tau).exp();
    let mut ratio = state.ratio + (target - state.ratio) * k;
    if (target - ratio).abs() < 0.0005 {
        ratio = target;
    }
    let mut s = state.clone();
    s.ratio = ratio;
    s
}

pub fn step(
    state: &CurtainState,
    cursor_x: Option<f64>,
    screen_w: f64,
    cfg: &CurtainConfig,
    now: i64,
    dt_ms: f64,
) -> CurtainState {
    let decided = decide(state, cursor_x, screen_w, now);
    advance(&decided, cfg, dt_ms)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_config() {
        let bad = CurtainConfig {
            collapsed: -0.5,
            expanded: 2.0,
        };
        let norm = normalize_config(Some(bad));
        assert_eq!(norm.collapsed, curtain_consts::MIN_COLLAPSED);
        assert_eq!(norm.expanded, curtain_consts::MAX_EXPANDED);
    }

    #[test]
    fn test_is_over_panel() {
        // écran 1000px, ratio 0.2 -> frontière à 800px
        assert!(is_over_panel(Some(850.0), 0.2, 1000.0));
        assert!(!is_over_panel(Some(750.0), 0.2, 1000.0));
        assert!(!is_over_panel(None, 0.2, 1000.0));
    }
}
